core/services/ocr2/plugins/ocr2keeper/evm21/registry.go

```diff
@@ -926,3 +936,82 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
+	txr, err := r.client.TransactionReceipt(r.ctx, txHash)
+	if err != nil {
+		return nil, common.Hash{}, err
+	}
+
+	return txr.BlockNumber, txr.BlockHash, nil
+}
+
+func (r *EvmRegistry) getCheckResult(p ocr2keepers.UpkeepPayload, reason, state uint8, retryable bool) ocr2keepers.CheckResult {
+	return ocr2keepers.CheckResult{
+		IneligibilityReason:    reason,
+		PipelineExecutionState: state,
+		Retryable:              retryable,
+		UpkeepID:               p.UpkeepID,
+		Trigger:                p.Trigger,
+		WorkID:                 p.WorkID,
+		FastGasWei:             big.NewInt(0),
+		LinkNative:             big.NewInt(0),
+	}
+}
+
+// verifyCheckBlock checks that the check block and hash are valid, returns the pipeline execution state and retryable
+func (r *EvmRegistry) verifyCheckBlock(ctx context.Context, checkBlock, upkeepId *big.Int, checkHash common.Hash) (state uint8, retryable bool) {
+	// verify check block number is not too old
+	if r.bs.latestBlock.Load()-checkBlock.Int64() > validCheckBlockRange {
+		r.lggr.Warnf("latest block is %d, check block number %s is too old for upkeepId %s", r.bs.latestBlock.Load(), checkBlock, upkeepId)
+		return CheckBlockTooOld, false
+	}
+
+	var h string
+	var ok bool
+	// verify check block number and hash are valid
+	h, ok = r.bs.queryBlocksMap(checkBlock.Int64())
+	if !ok {
+		r.lggr.Warnf("check block %s does not exist in block subscriber for upkeepId %s, querying eth client", checkBlock, upkeepId)
+		b, err := r.client.BlockByNumber(ctx, checkBlock)
+		if err != nil {
+			r.lggr.Warnf("failed to query block %s: %s", checkBlock, err.Error())
+			return CheckBlockInvalid, true
+		}
+		h = b.Hash().Hex()
+	}
+	if checkHash.Hex() != h {
+		r.lggr.Warnf("check block %s hash do not match. %s from block subscriber vs %s from trigger for upkeepId %s", checkBlock, h, checkHash.Hex(), upkeepId)
+		return CheckBlockInvalid, false
+	}
+	return NoPipelineError, false
+}
+
+// verifyLogExists checks that the log still exists on chain, returns failure reason and retryable
+func (r *EvmRegistry) verifyLogExists(upkeepId *big.Int, p ocr2keepers.UpkeepPayload) (uint8, bool) {
+	logBlockNumber := int64(p.Trigger.LogTriggerExtension.BlockNumber)
+	logBlockHash := common.BytesToHash(p.Trigger.LogTriggerExtension.BlockHash[:])
+	// if log block number is populated, check log block number and block hash
+	if logBlockNumber != 0 {
+		h, ok := r.bs.queryBlocksMap(logBlockNumber)
+		if ok && h == logBlockHash.Hex() {
+			r.lggr.Debugf("tx hash %s exists on chain at block number %d for upkeepId %s", hexutil.Encode(p.Trigger.LogTriggerExtension.TxHash[:]), logBlockNumber, upkeepId)
+			return UpkeepFailureReasonNone, false
+		}
+		r.lggr.Warnf("log block %d does not exist in block subscriber for upkeepId %s, querying eth client", logBlockNumber, upkeepId)
+	} else {
+		r.lggr.Warnf("log block not provided, querying eth client for tx hash %s for upkeepId %s", hexutil.Encode(p.Trigger.LogTriggerExtension.TxHash[:]), upkeepId)
```

Comment:
nit. i think we should log this without the else bracket too, i.e. also when we were not able to get logBlockNumber within bs

Reply:
you mean log this outside else?